File: HCS_Toolbox_ver02/lo2ho.m
function [L0, G0, nu_min, nu_max]=lo2ho(A,C,L)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% The function [L0, G0, nu_min, nu_max]=lpc2hpc(A,B,K) upgrades 
%% the linear observer (LO) 
%%
%%    dz/dt=A*z+f+L*(C*z-y) 
%%
%% to Homogeneous Observer (HO):     
%%                                                                      
%%   dz/dt=A*z+f+(L0+|Cz-y|^(nu-1)*d(log(|Cz-y|))*(L-L0))*(Cz-y)            
%%                                                                       
%%
%% The input parameters:                                             
%%      A  - system matrix (n x n)                                       
%%      C  - outout matrix (k x n)       
%%      L  - gain matrix (n x k) of LO s.t. A+L*C is Hurwitz  
%%
%% The output parameters:                                             
%%      L0 - gain (m x n) of linear component of HPC                                          
%%      G0 - matrix (n x n) that  defines the generator Gd=eye(n)+nu*G0  
%%           of the dilation d(s)=expm(s*Gd)                     
%%  nu_min - minimal admissible degree of HO (nu_min<=nu)  
%%  nu_max - maximal admissible degree of HO  (nu<=nu_max)
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
L0=0;  G0=0; nu_min=0;nu_max=0; 
tol=0.00001; 

sA=size(A); sC=size(C);

if sA(1)~=sA(2) disp('Error: the system matrix must be square'); return;
end;
if sC(2)~=sA(2) disp('Error:  dimensions of system  and output matrices must agree');return;
end;

[L0, G0, P, nu_max, nu_min]=lpc2hpc(A',C',L');
L0=L0'; G0=-G0'; nu_min=-nu_min;nu_max=-nu_max;

File: hcs_toolbox_py/lo2ho.py
"""
Upgrade Linear Observer (LO) to Homogeneous Observer (HO).

Original MATLAB: lo2ho.m

The upgrade uses duality: lo2ho(A,C,L) ≡ lpc2hpc(A',C',L') transposed.

    Linear:  dz/dt = A*z + f + L*(C*z - y)
    HO:      dz/dt = A*z + f + (L0 + |Cz-y|^(nu-1) * d(log|Cz-y|) * (L-L0)) * (Cz-y)
"""

import numpy as np
from .lpc2hpc import lpc2hpc


def lo2ho(A, C, L):
    """
    Upgrade a linear Luenberger observer to a Homogeneous Observer.

    Parameters
    ----------
    A : (n, n) ndarray
        System matrix.
    C : (k, n) ndarray
        Output matrix (full row rank).
    L : (n, k) ndarray
        Luenberger observer gain matrix. A + L*C must be Hurwitz.

    Returns
    -------
    L0 : (n, k) ndarray
        Gain of the linear component of HO.
    G0 : (n, n) ndarray
        Matrix defining the generator of dilation: Gd = I + nu*G0.
    nu_min : float
        Minimal admissible homogeneity degree (nu_min <= nu).
    nu_max : float
        Maximal admissible homogeneity degree (nu <= nu_max).
    """
    K0, G0, P, nu_max, nu_min = lpc2hpc(A.T, C.T, L.T)
    return K0.T, -G0.T, P, -nu_min, -nu_max
